File: C++/CreationalDesignPattern/BuilderDesign.cpp
#include <string>

class Student {
public:
    class StudentBuilder {
    public:
        std::string name;  // mandatory
        // optional fields
        int rollNumber = 11;
        long age = 12;
        int grade = 12;
        std::string phoneNumber = "dadsadasd";

        explicit StudentBuilder(const std::string& name) : name(name) {}

        StudentBuilder& setRollNumber(int rollNumber) {
            this->rollNumber = rollNumber;
            return *this;
        }

        StudentBuilder& setAge(long age) {
            this->age = age;
            return *this;
        }

        StudentBuilder& setGrade(int grade) {
            this->grade = grade;
            return *this;
        }

        StudentBuilder& setPhoneNumber(const std::string& phoneNumber) {
            this->phoneNumber = phoneNumber;
            return *this;
        }

        Student build() {
            return Student(*this);
        }
    };

    explicit Student(const StudentBuilder& builder)
        : name(builder.name),
          rollNumber(builder.rollNumber),
          age(builder.age),
          grade(builder.grade),
          phoneNumber(builder.phoneNumber) {}

private:
    std::string name;  // mandatory
    int rollNumber;
    long age;
    int grade;
    std::string phoneNumber;
};

int main() {
    Student student = Student::StudentBuilder("pradeep")
                           .setGrade(2)
                           .setPhoneNumber("padsad")
                           .build();

    return 0;
}

File: Python/CreationalDesignPattern/BuilderDesign.py
class Student:
    def __init__(self, builder: "Student.StudentBuilder"):
        self.name = builder.name
        self.roll_number = builder.roll_number
        self.age = builder.age
        self.grade = builder.grade
        self.phone_number = builder.phone_number

    class StudentBuilder:
        def __init__(self, name: str):
            self.name = name  # mandatory
            # optional fields
            self.roll_number = 11
            self.age = 12
            self.grade = 12
            self.phone_number = "dadsadasd"

        def set_roll_number(self, roll_number: int) -> "Student.StudentBuilder":
            self.roll_number = roll_number
            return self

        def set_age(self, age: int) -> "Student.StudentBuilder":
            self.age = age
            return self

        def set_grade(self, grade: int) -> "Student.StudentBuilder":
            self.grade = grade
            return self

        def set_phone_number(self, phone_number: str) -> "Student.StudentBuilder":
            self.phone_number = phone_number
            return self

        def build(self) -> "Student":
            return Student(self)


def main():
    student = (
        Student.StudentBuilder("pradeep")
        .set_grade(2)
        .set_phone_number("padsad")
        .build()
    )


if __name__ == "__main__":
    main()
